### PR TITLE
https://issuehunt.io/repos/114969039/issues/8

### PR DESCRIPTION
**Pull request by june07**
Added Google Cloud Storage support

- Upload backup data to Google Cloud Storage
- Restore from Google Cloud Storage
- Manage data on Google Cloud Storage
- Delete expired files
- Add environment variables to configure
- Modify README.md

All items have been completed.

### CODE REVIEW

File: bin/backup.sh [hunk 1]
```diff
@@ -32,11 +33,16 @@ TARBALL_FULLPATH="${TMPDIR}/${TARBALL}"
 
 
 # check parameters
-if [ "x${S3_TARGET_BUCKET_URL}" == "x" ]; then
-  echo "ERROR: The environment variable S3_TARGET_BUCKET_URL must be specified." 1>&2
+# deprecate the old option
+if [ "x${S3_TARGET_BUCKET_URL}" != "x" ]; then
+  TARGET_BUCKET_URL=$S3_TARGET_BUCKET_URL
```
Comment (ryu-sato) [inline bin/backup.sh:39]: 👍 
Could you add print message like "WARNING: The environment variable S3_TARGET_BUCKET_URL is deprecated." to announce to users.

File: README.md [hunk 2]
```diff
@@ -62,7 +73,10 @@ You can use "**restore**" command to restore database from backup file.
 docker run --rm \
   -e AWS_ACCESS_KEY_ID=<Your IAM Access Key ID> \
   -e AWS_SECRET_ACCESS_KEY=<Your IAM Secret Access Key> \
-  -e S3_TARGET_BUCKET_URL=<Target S3 Bucket URL (s3://...)> \
+  -e GCP_ACCESS_KEY_ID=<Your GCP Access Key> \
+  -e GCP_SECRET_ACCESS_KEY=<Your GCP Secret> \
+  -e GCP_PROJECT_ID=<Your GCP Project ID> \
+  -e TARGET_BUCKET_URL=<Target Bucket URL ([s3://...|gs://...])> \
   -e S3_TARGET_FILE=<Target S3 file name to restore> \
```
Comment (ryu-sato) [inline README.md]: `S3_TARGET_FILE` seem to be changed to `TARGET_FILE`, right?
Please modify this.

File: test/docker-compose.yml [hunk 3]
```diff
@@ -27,7 +27,7 @@ services:
     <<: *app_default
     environment:
       <<: *base_env
-      S3_TARGET_BUCKET_URL: s3://app_restore/
+      TARGET_BUCKET_URL: s3://app_restore/
       S3_TARGET_FILE: dummy-backup-20180622.tar.bz2
```
Comment (ryu-sato) [inline test/docker-compose.yml]: Please modify `S3_TARGET_FILE` to `TARGET_FILE`.

**Review by ryu-sato — CHANGES_REQUESTED** [review]
1. Please modify `README.md`, `bin/backup.sh`, and ` test/docker-compose.yml`. I was commented the point of modification.
2. The PR is implemented to do with HMAC authentication, but GCS prefered authentication with OAuth 2. ([ref](https://cloud.google.com/storage/docs/gsutil/addlhelp/CredentialTypesSupportingVariousUseCases))
Most of the tasks of the issue are completed, but could you implement with OAuth2?
Maybe the option of `docker run` will differ from S3 pattern like below.


```
$ docker run -it --net growidockercompose_default --rm -e TARGET_BUCKET_URL=gs://***/ weseek/mongodb-awesome-backup
This command will create a boto config file at /root/.boto containing
your credentials, based on your responses to the following questions.
Please navigate your browser to the following URL:
https://accounts.google.com/o/oauth2/auth?scope=***
In your browser you should see a page that requests you to authorize access to Google Cloud Platform APIs and Services on your behalf. After you approve, an authorization code will be displayed.

Enter the authorization code: **************

Please navigate your browser to https://cloud.google.com/console#/project,
then find the project you will use, and copy the Project ID string from the
second column. Older projects do not have Project ID strings. For such projects,
 click the project and then copy the Project Number listed under that project.

What is your project-id? ryutest-99285

gsutil developers rely on user feedback to make improvements to the
tool. Would you like to send anonymous usage statistics to help
improve gsutil? [y/N]

Boto config file "/root/.boto" created. If you need to use a proxy to
access the Internet please see the instructions in that file.
=== ./backup.sh started at 2019/02/19 19:26:56 ===
dump MongoDB...
2019-02-19T19:26:56.590+0900    writing admin.system.version to
  : <snip>
```

**Comment by ryu-sato** [comment]
> 2. The PR is implemented to do with HMAC authentication, but GCS prefered authentication with OAuth 2. ([ref](https://cloud.google.com/storage/docs/gsutil/addlhelp/CredentialTypesSupportingVariousUseCases))
>     Most of the tasks of the issue are completed, but could you implement with OAuth2?
>     Maybe the option of `docker run` will differ from S3 pattern like below.
> ```
> $ docker run -it --net growidockercompose_default --rm -e TARGET_BUCKET_URL=gs://***/ weseek/mongodb-awesome-backup
> ```

Above example was not good.

We thought better example is that user create `.boto` by `gsutil` at first, and next execute docker container with mounting `.boto` file like below.

```
$ cd /path/to/mongodb-awesome-backup-repository/
$ gsutil config -o gsutil-config/.boto # Create .boto file at first. (*1)
$ docker run -it  --net growidockercompose_default -v $(pwd)/gsutil-conf/.boto:/root/.boto --rm -v config/.boto:/root/.boto -e TARGET_BUCKET_URL=gs://***/ weseek/mongodb-awesome-backup
=== ./backup.sh started at 2019/02/22 18:59:06 ===
dump MongoDB...
  : <snip>
```
(*1) It is needed to install gsutil in host of docker container.

Finally we expected to **REMAIN** implementation with HMAC authentication.

File: README.md (hunk 2 above)
Comment (june07) [reply README.md]: Done.

File: bin/backup.sh (hunk 1 above)
Comment (june07) [reply bin/backup.sh:39]: Done.

File: test/docker-compose.yml (hunk 3 above)
Comment (june07) [reply test/docker-compose.yml]: Done.

**Comment by june07** [comment]
The reason I chose to go with HMAC authentication was to maintain parity with the current AWS S3 implementation.  Although oauth is the recommended method, it requires input from the user at runtime which is different from the current behavior of the S3 code.  If oauth is used then the user will be prompted to provide input...

**Comment by ryu-sato** [comment]
> 1. Please modify `README.md`, `bin/backup.sh`, and ` test/docker-compose.yml`. I was commented the point of modification.

👍 OK.

> The reason I chose to go with HMAC authentication was to maintain parity with the current AWS S3 implementation. Although oauth is the recommended method, it requires input from the user at runtime which is different from the current behavior of the S3 code. If oauth is used then the user will be prompted to provide input...

Yes...

I think implementing GCS with HMAC authentication is a good simple idea.
But enable to OAuth is a better idea.

I think all we needs to enable OAuth is that execute `boto.sh` when `/root/.boto` is not exists.
So, I think that it will not force a heavy load to user and to you.

**Comment by june07** [comment]
Added auth0 support and updated the README to reflect the changes.  Works nicely in that user input is only required on the first run and the .boto is saved to the user home directory (can be changed easily).

File: README.md [hunk 4]
```diff
@@ -12,20 +12,31 @@ Requirements
 * Amazon IAM Access Key ID/Secret Access Key
   * which must have the access lights of the target Amazon S3 bucket.
 
+OR
+
+* Google Cloud Interoperable storage access keys (see https://cloud.google.com/storage/docs/migrating#keys)
+  * GCP_ACCESS_KEY_ID, GCP_SECRET_ACCESS_KEY, and GCP_PROJECT_ID are only required if using HMAC authentication.
+  * When using oauth authentication, a docker mount ``` -v ~:/mab``` can be added to save auth0 credentials to your home directory after mongodb-awesome-backup is run.  On subsequent runs, the same ```~/.bota``` file will be used for authentication.
```
Comment (skomma) [inline README.md]: `` ``` `` may be `` ` ``?

And, what are `mab` and `bota`?
Are there strings samples?

File: README.md [hunk 5]
```diff
@@ -12,20 +12,31 @@ Requirements
...
+Note that either AWS_ or GCP_ vars are required not both.
 
 ```bash
 docker run --rm \
   -e AWS_ACCESS_KEY_ID=<Your IAM Access Key ID> \
   -e AWS_SECRET_ACCESS_KEY=<Your IAM Secret Access Key> \
-  -e S3_TARGET_BUCKET_URL=<Target S3 Bucket URL (s3://...)> \
+  [ -e GCP_ACCESS_KEY_ID=<Your GCP Access Key> \
+  -e GCP_SECRET_ACCESS_KEY=<Your GCP Secret> \
+  -e GCP_PROJECT_ID=<Your GCP Project ID> \ ]
```
Comment (skomma) [inline README.md]: Could you add `[]` in each line?

File: README.md [hunk 6]
```diff
@@ -42,15 +53,19 @@ and after running this, `backup-YYYYMMdd.tar.bz2` will be placed on Target S3 Bu
...
+  -e TARGET_BUCKET_URL=<Target Bucket URL ([s3://...|gs://...])> \
```
Comment (skomma) [inline README.md:60]: Could you add `[]` in each line?

File: README.md [hunk 7]
```diff
@@ -62,14 +77,18 @@ You can use "**restore**" command to restore database from backup file.
 docker run --rm \
   -e AWS_ACCESS_KEY_ID=<Your IAM Access Key ID> \
   -e AWS_SECRET_ACCESS_KEY=<Your IAM Secret Access Key> \
-  -e S3_TARGET_BUCKET_URL=<Target S3 Bucket URL (s3://...)> \
-  -e S3_TARGET_FILE=<Target S3 file name to restore> \
+  [ -e GCP_ACCESS_KEY_ID=<Your GCP Access Key> \
+  -e GCP_SECRET_ACCESS_KEY=<Your GCP Secret> \
+  -e GCP_PROJECT_ID=<Your GCP Project ID> \ ]
+  -e TARGET_BUCKET_URL=<Target Bucket URL ([s3://...|gs://...])> \
+  -e TARGET_FILE=<Target S3 or GS file name to restore> \
```
Comment (skomma) [inline README.md:85]: Could you add `[]` in each line?

**Review by skomma — CHANGES_REQUESTED** [review]
I have a few questions, and could you fix a few points?

File: README.md (hunk 7 above)
Comment (june07) [reply README.md:85]: Added.

File: README.md (hunk 4 above)
Comment (june07) [reply README.md]: Replaced ``` with `.  "mab" is the name of the mount point inside of the Docker container, pretty standard stuff.

File: README.md [hunk 8]
```diff
@@ -12,20 +12,31 @@ Requirements
 * Amazon IAM Access Key ID/Secret Access Key
   * which must have the access lights of the target Amazon S3 bucket.
 
+OR
+
+* Google Cloud Interoperable storage access keys (see https://cloud.google.com/storage/docs/migrating#keys)
+  * GCP_ACCESS_KEY_ID, GCP_SECRET_ACCESS_KEY, and GCP_PROJECT_ID are only required if using HMAC authentication.
+  * When using oauth authentication, a docker mount ` -v ~:/mab` can be added to save auth0 credentials to your home directory after mongodb-awesome-backup is run.  On subsequent runs, the same `~/.bota` file will be used for authentication.
```
Comment (skomma) [inline README.md]: Perhaps, is `mab` the abbreviation of "Mongodb Awesome Backup"?

And is `~/.bota` mistake of `~/.boto`??

**Review by skomma — CHANGES_REQUESTED** [review]

**Review by ryu-sato — APPROVED** [review]
> Added auth0 support and updated the README to reflect the changes. Works nicely in that user input is only required on the first run and the .boto is saved to the user home directory (can be changed easily).

Good Job! approve.

File: README.md (hunk 8 above)
Comment (june07) [reply README.md]: Ha, ha!  Yes, you are right... bota should be boto!  Fixed it.  I did that more than once during the execution of this PR.  As you're 2 for 2, as mab is the abbreviation of "Mongodb Awesome Backup".  Cheers!

Comment (skomma) [reply README.md]: Hmm...I think users cannot read the meaning of "mab" from this document like me...

Could you add an explanation of "mab"?
This is the final change request.

Comment (june07) [reply README.md]: Done

**Review by skomma — APPROVED** [review]

**Pull request merged**